File: frontend/src/Components/Popups/CreateMusicAlbumPopup.js
import React, { useState } from 'react';
import server from '../Static/Constants';

const CreateMusicAlbumPopup = ({ onClose }) => {
    const [spotifyLink, setSpotifyLink] = useState("");
    const [appleLink, setAppleLink] = useState("");
    const [errorMessage, setErrorMessage] = useState(null);
    const [successMessage, setSuccessMessage] = useState(null);

    const handleCreateAlbum = async () => {
        setErrorMessage(null);
        setSuccessMessage(null);

        if (!spotifyLink || !appleLink) {
            setErrorMessage("Please fill in all fields.");
            return;
        }

        try {
            const token = localStorage.getItem("token");
            const response = await fetch(`${server}/upload/music_links/`, {
                method: "POST",
                headers: {
                    "Authorization": token,
                    "Content-Type": "application/json"
                },
                body: JSON.stringify({
                    spotify_link: spotifyLink,
                    apple_link: appleLink,
                })
            });

            if (response.ok) {
                setSuccessMessage("Album created successfully! Album is being processed.");
                setSpotifyLink("");
                setAppleLink("");

            } else {
                const errorData = await response.json();
                setErrorMessage(errorData.message || "An error occurred during album creation.");
            }
        } catch (error) {
            console.error("Error creating album:", error);
            setErrorMessage("An error occurred during album creation.");
        }
    };

    return (
        <div className="popup-overlay">
            <div className="popup-content">
                <h2>Create Music Album</h2>

                {errorMessage && <p className="error-message" style={{ color: "red" }}>{errorMessage}</p>}
                {successMessage && <p className="success-message" style={{ color: "green" }}>{successMessage}</p>}

                <input
                    type="text"
                    placeholder="Album Spotify Link"
                    value={spotifyLink}
                    onChange={(e) => setSpotifyLink(e.target.value)}
                />
                <textarea
                    placeholder="Album Apple Music Link"
                    value={appleLink}
                    onChange={(e) => setAppleLink(e.target.value)}
                />
                <button onClick={handleCreateAlbum}>Create Album</button>
                <button onClick={onClose}>Cancel</button>
            </div>
        </div>
    );
};

export default CreateMusicAlbumPopup;

File: flutter_frontend/lib/assets/music/popups/createMusicAlbumPopup.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:flutter_frontend/static/constants.dart';

class CreateMusicAlbumPopup extends StatefulWidget {
  const CreateMusicAlbumPopup({Key? key}) : super(key: key);

  @override
  State<CreateMusicAlbumPopup> createState() => _CreateMusicAlbumPopupState();
}

class _CreateMusicAlbumPopupState extends State<CreateMusicAlbumPopup> {
  final FlutterSecureStorage _storage = const FlutterSecureStorage();

  final TextEditingController _spotifyController = TextEditingController();
  final TextEditingController _appleController = TextEditingController();

  bool _isCreating = false;
  String? _errorMessage;
  String? _successMessage;

  Future<void> _createAlbum() async {
    setState(() {
      _isCreating = true;
      _errorMessage = null;
      _successMessage = null;
    });

    if (_spotifyController.text.isEmpty || _appleController.text.isEmpty) {
      setState(() {
        _errorMessage = 'Please fill in all fields.';
        _isCreating = false;
      });
      return;
    }

    try {
      final token = await _storage.read(key: 'token');

      final response = await http.post(
        Uri.parse('$server/upload/music_links/'),
        headers: {
          'Authorization': token ?? '',
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'spotify_link': _spotifyController.text.trim(),
          'apple_link': _appleController.text.trim(),
        }),
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        setState(() {
          _successMessage =
              'Album created successfully! Album is being processed.';
          _spotifyController.clear();
          _appleController.clear();
          _isCreating = false;
        });
      } else {
        final data = jsonDecode(response.body);
        setState(() {
          _errorMessage =
              data['message'] ?? 'An error occurred during album creation.';
          _isCreating = false;
        });
      }
    } catch (e) {
      setState(() {
        _errorMessage = 'An error occurred during album creation.';
        _isCreating = false;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return AlertDialog(
      title: const Text('Create Music Album'),
      content: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          if (_errorMessage != null)
            Padding(
              padding: const EdgeInsets.only(bottom: 8),
              child: Text(
                _errorMessage!,
                style: const TextStyle(color: Colors.red),
              ),
            ),
          if (_successMessage != null)
            Padding(
              padding: const EdgeInsets.only(bottom: 8),
              child: Text(
                _successMessage!,
                style: const TextStyle(color: Colors.green),
              ),
            ),
          TextField(
            controller: _spotifyController,
            decoration: const InputDecoration(labelText: 'Album Spotify Link'),
          ),
          const SizedBox(height: 10),
          TextField(
            controller: _appleController,
            decoration: const InputDecoration(
              labelText: 'Album Apple Music Link',
            ),
            maxLines: 2,
          ),
        ],
      ),
      actions: [
        TextButton(
          onPressed: _isCreating ? null : () => Navigator.of(context).pop(),
          child: const Text('Cancel'),
        ),
        ElevatedButton(
          onPressed: _isCreating ? null : _createAlbum,
          child: _isCreating
              ? const SizedBox(
                  height: 18,
                  width: 18,
                  child: CircularProgressIndicator(strokeWidth: 2),
                )
              : const Text('Create Album'),
        ),
      ],
    );
  }
}
